Ghost constraints

vl0w opened the pull request:
Adds ghost constraints to Prusti, which we plan to use for `Iterator` support in Prusti.

A ghost constraint defines a constraint for a specification which must hold for a function call.
That means, the specification is only "active" for a function call when the constraint is fulfilled on callsite.
At this point, a ghost constraint is simply a trait bound for a type (usually a generic / `Self`). More technically, a ghost constraint is conceptually a Rust `where` clause attached to a spec.

```rust
#[ghost_constraint(T: SomeTrait, [
    requires(R), ensures(E)
])]
fn foo<T>(...) {...}

fn client() {
    // Assume T1 implements SomeTrait
    // The precondition R and the postcondition E of foo are active
    foo::<T1>(...); 

    // Assume T2 does not implement SomeTrait
    // The precondition R and the postcondition E of foo are _NOT_ active
    foo::<T2>(...); 
}

```

Conceptually, a contract of a function is split into a specification without constraints and several specifications which have constraints (constrained specs). A constrained spec is **active** for a function call when it's bounds are satisfied. If no constrained spec is active, the base spec is active. Only one spec may be active at a time. This PR should be restrictive enough such that multiple active specs are not possible.
The constrained specs gain knowledge about their bounds, so the traits in the bound can be used in the specification.

## Goal
The goal of this PR is to enable modular specifications on iterators and iterator adapters.
A draft specification outline is the following:

```rust
// Defines two abstract predicates
trait IteratorSpecExt {
    fn enumerated(&self) -> bool;
    fn completed(&self) -> bool;
}


#[extern_spec]
trait Iterator {
    // Abstract predicates used in the spec
    // of Iterator
    #[requires(false)] // Self without IteratorSpecExt is uncallable
    #[ghost_constraint(Self: IteratorSpecExt, [
        requires(self.enumerated()),
        ensures(self.enumerated())
    ])]
    fn next(&mut self) -> Option<Self::Item>;
}

impl<...> IteratorSpecExt for std::iter::Map<...> {
    // Give a meaningful implementation to the abstract predicates
}

#[extern_spec]
impl<...> Iterator for std::iter::Map<...> {

    // More specs
    fn next(&mut self) -> Option<Self::Item>;
}

fn client(iter: std::iter::Map<...>) {
    iter.next();
    // iter.enumerated() holds
}

```

## Challenge
Checking correct behavioral subtyping rules involving ghost constraints is rather involved because a contract may contain arbitrarily constrained specs with overlapping constraints. 
Consider a a user-typed contract with a base spec `BS` and a set of constrained specs `{CS_1, CS_2, ..., CS_n}`, i.e.
```rust
#[requires(pre(BS))]
#[ensures(post(BS))]
#[ghost_constraint(bounds(CS_1), [requires(pre(CS_1)), ensures(post(CS_1))])]
// ...
#[ghost_constraint(bounds(CS_n), [requires(pre(CS_n)), ensures(post(CS_n))])]
fn foo<T>(...) {...}
```
For example:
```rust
#[requires(requires(...))]
#[ensures(ensures(...))]
#[ghost_constraint(T: A, [...])]
#[ghost_constraint(T: B + C, [...])]
#[ghost_constraint(T: A + C, [...])]
fn foo<T>(...) {...}
```
Generally, we face the following challenges:
- What is the notion of subtyping amongst `BS` and `CS_1, ..., CS_n`
- Which spec should be active on a callsite?
- How should `foo<T>` be modularly verified?

Since this PR is intended to enable modular verification support for iterators only,
this PR does not fully resolve these challenges.

## Noteworthy implementation details
### Refinement is not automatically verified
- There is no check that every `CS_i` is a valid refinement of `BS`
- There is no check that `CS_i` is a valid refinement of `CS_j` when the bounds defined on `CS_i` induce a subtyping check w.r.t. the bounds of `CS_j`

The Prusti user needs to make sure that no soundness issues are introduced when using constrained specs. This is one of the main reasons why we add a feature flag for ghost constraints.

### Only allowed in `#[trusted]` functions (including extern specs)
For a modular verification, it is not immediately obvious which spec of a contract is active for any call to that function. This PR is thus very restrictive and allows constrained specs only on trusted functions (and external specs).

### It is not possible to declare multiple ghost constraints with different bounds
This is ensured during constraint resolving of a function contract. It basically checks that `bounds(CS_i) == bounds(CS_j)` for every `i != j`.
The following spec will be rejected by Prusti:
```rust
#[ghost_constraint(T: A, [...])]
#[ghost_constraint(T: B, [...])]
fn foo<T>() { }
```
The following is possible however: 
```rust
#[ghost_constraint(T: A, [...])]
#[ghost_constraint(T: A, [...])]
fn foo<T>() { }
```
Which will unify both ghost constraints.

### Inheriting from the base spec to the constrained spec
A constrained spec "inherits" the postconditions, pure flag, trusted flag, pledges from the base spec. This is mainly a convenience feature to not need to repeat the specs.

Preconditions are not copied, since this would not lead to a valid refinement.

- [ ] Maybe pledges should not be copied @Aurel300 ?

### Trait contract refinement
Trait contract refinement is aware of ghost constraints to account for ghost constraints on the trait method with a `Self` bound.

Given a contract `T_BS, T_CS_1` on a trait and `S_BS, S_CS_1`, contract refinement for a function call is performed like this:
- Get the contract of the impl contract
- Resolve the obligations of the impl 
    - -> Selected spec `X`
- Get the contract of the trait
- Resolve the obligations of the trait contract
    - -> Selected spec `Y`
- Refine `X` with `Y` -> Refined `Z`
- `Z` is the spec used for encoding a function call
- `X` and `Y` are used for contract refinement checks

Note 1: There cannot be multiple ghost constraints on either the trait or the impl.

Note 2: Obligation resolving does not happens not pure/trusted checks (we still refine though)

Note 3: This refinement process happens once for the encoding and modular verification of the function _and_ once for any encoding of a call. Specifically, this does refinement does not happen for any outcome of obligation resolving. We might need to adjust this in a future PR.

## Work to do
### Parsing
- [x] Support generic methods in trait bounds
- [x] Macro can not be parsed in certain cases (see corresponding TODO in code)
- [x] Trait/Impls -> Disallow usage in refined trait specs
- [x] Further parsing unit tests
- [x] Extern specs
- [x] Fully qualified paths in trait bounds

### Architectural
- [x] Reiterate structure in `typed.rs`
- [x] Naming: Ghost constraint vs obligation (in struct names etc)
- [x] Associated types in trait bounds (lead to an error)

### General
- [x] Resolve all  `TODO hansenj: ...`
- [x] Disallow multiple ghost constraints
- [x] ~User Docs~
- [x] `Self` in ghost constraints (and extern specs)
- [x] Add feature flag
- [x] Explain/Document ghost constraints in the context of refinements

vl0w committed:
Initial support for ghost constraints

vl0w committed:
Ghost constraint parsing (refactoring, tests)

vl0w committed:
Fix parsing of proc macro with multiple parameters

vl0w committed:
Disallow obligation refinement

- Includes refactorings
- Store obligations along proc specification
- Disallow/disable refinement of obligations

Aurel300 commented on prusti-interface/src/environment/mod.rs:
```diff
@@ -389,6 +389,7 @@ impl<'tcx> Environment<'tcx> {
     /// Checks whether the given type implements the trait with the given DefId.
     /// Accounts for generic params on the trait by the given substs.
     /// TODO hansenj: name, params, docs
+    /// TODO hansenj: Do we need this?
```
I assume yes? A ghost constraint `T: Foo<i32>` is not the same as `T: Foo<bool>`.

vl0w committed:
Add comment to ghost_constraint macro

vl0w committed:
Better errors parsing ghost_constraint macro

vl0w committed:
Add test (currently ignored)

vl0w committed:
Allow ghost constraints only in trusted functions + inherit specs from base contract

vl0w committed:
Move obligations out of ProcedureSpecification

vl0w committed:
Remove SpecificationSet

vl0w committed:
Improved logging in constraints.rs

vl0w committed:
Support ghost constraints on `Self` in extern trait specs

vl0w committed:
Add more parsing unit tests

vl0w committed:
Refactoring/implementation of constraint resolving

- Simplified interface of specs module
- Solved bugs which came up during refinement

vl0w committed:
Do not copy preconditions from base spec when collecting constrained specs

vl0w committed:
Check trustedness of function when adding ghost constraints during collection

vl0w committed:
Renamed test folder and tests

vl0w committed:
Remove superfluous TODO

vl0w committed:
Add comments / resolve TODO

vl0w committed:
Add feature flag for ghost constraints

vl0w committed:
Rename SpecsWithConstraints to SpecGraph

vl0w committed:
Adjust refinement caching

The refinement is now cached per SpecQuery

Aurel300 commented on prusti-common/src/config.rs:
```diff
@@ -533,3 +534,6 @@ pub fn full_compilation() -> bool {
 }
+
+/// Enables ghost constraint parsing and resolving.
+pub fn enable_ghost_constraints() -> bool { read_setting("enable_ghost_constraints") }
```
```suggestion
pub fn enable_ghost_constraints() -> bool {
    read_setting("enable_ghost_constraints")
}
```

Aurel300 commented on docs/dev-guide/src/config/flags.md:
```diff
@@ -105,6 +106,10 @@ When enabled, reborrowing DAGs will be output in debug files.
 
+## `ENABLE_GHOST_CONSTRAINTS`
+
+Enables ghost constraints in Prusti specifications.
```
Point out that this is an experimental feature.

Aurel300 commented on prusti-interface/src/environment/mod.rs:
```diff
@@ -495,10 +495,16 @@ impl<'tcx> Environment<'tcx> {
 
     /// Checks whether the given type implements the trait with the given DefId.
     pub fn type_implements_trait(&self, ty: ty::Ty<'tcx>, trait_def_id: DefId, param_env: ty::ParamEnv<'tcx>) -> bool {
```
This method should probably not exist: I assume the empty `ty::List` will cause a crash when checking against a generic trait.

vl0w replied:
Afaik this is not used in the `specifications` module or is used anyhow involved in constraint resolving (`constraints.rs`)?

Aurel300 commented on prusti-interface/src/specs/mod.rs:
```diff
@@ -71,65 +72,50 @@ impl<'a, 'tcx> SpecCollector<'a, 'tcx> {
                     }
                     SpecIdRef::Pledge { lhs, rhs } => {
-                        pledges.push(typed::Pledge {
+                        let pledge = typed::Pledge {
```
Any reason for this local variable? Seems inconsistent with the other cases.

Aurel300 commented on prusti-interface/src/specs/typed.rs:
```diff
@@ -1,68 +1,157 @@
+        match self.get_constraint(pre, env) {
+            None => {
+                self.base_spec.pres.push(pre);
+                // Preconditions are explicitly not copied (as opposed to postconditions)
+                // This would always violate behavioral subtyping rules
```
Maybe explain this in more detail (with an example) in a comment above `impl SpecGraph<ProcedureSpecification>`, and mention briefly that this does not happen here and *does* happen in postconditions.

vl0w replied:
I added a big comment to `SpecGraph`. Hope it clarifies things.

Aurel300 commented on prusti-interface/src/specs/typed.rs:
```diff
@@ -1,68 +1,157 @@
+            .or_insert_with(|| self.base_spec.clone())
+    }
+
+    /// Note: First wins, we do currently support not multiple constraints
```
```suggestion
    /// Note: First wins, we currently do not support multiple constraints
```

Aurel300 commented on prusti-interface/src/specs/typed.rs:
```diff
@@ -126,11 +215,55 @@ impl<T> SpecificationItem<T> {
         self.extract_with_strategy(|(_, refined)| refined)
     }
+
+    pub fn expect_empty_or_inherent(&self) -> Option<&T> {
```
These types of methods (all the `expect_*`) are nicer with `#[track_caller]`.

Aurel300 commented on prusti-specs/src/extern_spec_rewriter/common.rs:
```diff
@@ -141,11 +142,40 @@ impl<'a> AssociatedTypeRewriter<'a> {
+    fn rewrite_tokens(&self, self_replacement_as_str: &str, tokens: TokenStream) -> TokenStream {
+        let it = tokens.into_iter();
+        it.map(|tt| {
+            match tt{
```
```suggestion
            match tt {
```

Aurel300 commented on prusti-specs/src/extern_spec_rewriter/common.rs:
```diff
@@ -141,11 +142,40 @@ impl<'a> AssociatedTypeRewriter<'a> {
+        let self_replacement = self.self_type.to_token_stream().to_string();
+
+        // Note: Attribute tokens are not visited by a syn visitor,
+        // so we use visit the attribute tokens "manually"
```
```suggestion
        // so we visit the attribute tokens "manually"
```

Aurel300 commented on prusti-specs/src/lib.rs:
```diff
@@ -501,4 +515,4 @@ pub fn predicate(tokens: TokenStream) -> TokenStream {
         #cleaned_fn
     }
-}
+}
```
```suggestion
}

```

(and in general, please try to keep the line ends)

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -131,9 +139,12 @@ impl PrustiTokenStream {
                         pos += 1;
                         tokens.push_back(PrustiToken::Token(token.clone()));
-                        if p.spacing() != Joint {
+                        if p.spacing() != Alone {
```
I'm not sure I understand these changes. Why stop parsing when a `Joint`-spaced token is encountered?

vl0w replied:
I removed this and adjusted the tests (as discussed)

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -421,14 +452,15 @@ impl PrustiTokenStream {
         res
     }
 
+
```
```suggestion
```

Aurel300 commented on prusti-tests/tests/compiletest.rs:
```diff
@@ -126,6 +126,7 @@ fn run_verification_base(group_name: &str, filter: &Option<String>) {
         TemporaryEnvVar::set("PRUSTI_FULL_COMPILATION", "true"),
         TemporaryEnvVar::set("PRUSTI_ENCODE_UNSIGNED_NUM_CONSTRAINT", "true"),
         TemporaryEnvVar::set("PRUSTI_QUIET", "true"),
+        TemporaryEnvVar::set("PRUSTI_ENABLE_GHOST_CONSTRAINTS", "true"),
```
Let's only enable it per-test for now?

vl0w replied:
With the feature generally activated, we can check for regressions against code without ghost constraints.
Concretely, I would like to check that the specs module (w/ the changes in this PR) is still is correct.
I would therefore suggest that we keep the flag enabled. Eventually, ghost constraints won't be feature-gated anymore.

Aurel300 replied:
In the meanwhile though we are not testing any part of Prusti with the feature disabled, and it being disabled will be the default for any user. We can check for regressions when we make the feature enabled by default, not now. (And of course you can see locally and in the already-executed CI run if it caused any issues yet.)

Aurel300 commented on prusti-tests/tests/parse/fail/ghost-constraints/trait-bounds-illegal-4.rs:
```diff
@@ -0,0 +1,14 @@
+trait A { }
+trait B { }
+
+#[ghost_constraint(T: Fn(A) -> B, [ //~ ERROR: unexpected Prusti syntax
```
I'm not sure I understand why this should fail.

vl0w replied:
Ah good point. It should not fail but it currently does because the parser can't handle it.
I will remove the test.

Aurel300 commented on prusti-tests/tests/verify/fail/ghost-constraints/invalid-trait-refinement-1.rs:
```diff
@@ -0,0 +1,47 @@
+    fn foo(&mut self);
+}
+
+#[extern_spec] //~ ERROR: the method's precondition may not be a valid weakening of the trait's precondition
```
Why is it reported here?

vl0w replied:
The full error message is:
```
error: [Prusti: verification error] the method's precondition may not be a valid weakening of the trait's precondition.
  --> verify-me.rs:15:1
   |
15 | #[extern_spec] //~ ERROR: the method's precondition may not be a valid weakening of the trait's precondition
   | ^^^^^^^^^^^^^^
...
29 |     #[requires(self.x >= 15)]
   |                ^^^^^^^^^^^^
   |
   = help: The trait's precondition should imply the implemented method's precondition.
   = note: this error originates in the attribute macro `extern_spec` (in Nightly builds, run with -Z macro-backtrace for more info)
```
which is fine imo.
The `#[extern_spec]` is part of the error span because the precondition function is generated during the expansion of the extern spec.
I will create UI tests for `invalid-trait-refinement-1.rs` and `invalid-trait-refinement-2.rs` to resolve this confusion (afaik it's not possible to test a multiline span nicely with these error patterns).

Aurel300 commented on prusti-tests/tests/verify_overflow/fail/ghost-constraints/different-bounds.rs:
```diff
@@ -0,0 +1,23 @@
+// Note: For now, it is completely fine to _declare_ two ghost constraints with different bounds
+// since resolving happens for specific callsites. That is, without the call in main, this
+// file verifies.
```
This is not great but we're going to have multiple ghost constraints later anyway so that's fine.

vl0w replied:
Yeah, that's indeed a bit special with the current implementation. I guess when we'll implement behavioral subtyping checks of ghost constraints, we need to preprocess all of them somehow and catch these errors there.

Aurel300 commented on prusti-tests/tests/verify_overflow/pass/ghost-constraints/ghost-constraints-extend-base-attributes.rs:
```diff
@@ -0,0 +1,58 @@
+    let res = constrained_contract_inherits_posts_unsound(32 as i32);
+    assert!(res == 42);
+    assert!(res == 24);
+    assert!(false);
```
This is unfortunate but let's not actually test for asserting false (:

vl0w replied:
I believe it might be valuable to illustrate how one can introduce unsound verification results.
But yes, this is rather documentation than a test.

Aurel300 commented on prusti-tests/tests/verify_overflow/pass/ghost-constraints/splitted-constraint.rs:
```diff
@@ -0,0 +1,22 @@
+use prusti_contracts::*;
```
(name: `split-constraint.rs`?)

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4770,7 +4770,7 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
                     mir::Rvalue::Aggregate(box mir::AggregateKind::Closure(cl_def_id, cl_substs), _),
                 )) = stmt.kind {
-                    if let Some(spec) = self.encoder.get_loop_specs(cl_def_id) {
+                    if let Some(spec) = self.encoder.get_loop_specs(cl_def_id,  cl_substs) {
```
```suggestion
                    if let Some(spec) = self.encoder.get_loop_specs(cl_def_id, cl_substs) {
```

Aurel300 commented on prusti-viper/src/encoder/mir/specifications/interface.rs:
```diff
@@ -1,39 +1,98 @@
 
-    fn is_trusted(&self, def_id: DefId) -> bool;
+pub(crate) trait SpecificationsInterface<'tcx> {
+    fn is_pure(&self, def_id: DefId, substs: Option<SubstsRef<'tcx>>) -> bool;
```
Unsure how I feel about `substs` being passed for `is_pure` and `is_trusted`. These are flags which we cannot change with ghost constraints.

vl0w replied:
You are right, we cannot override `is_pure` and `is_trusted` with ghost constraints. We technically need `substs` for refinement though. Currently this is not handled properly as you mentioned in [this comment](https://github.com/viperproject/prusti-dev/pull/906#discussion_r844195348).

vl0w replied:
I indeed already use the correct method to resolve to the trait spec [here](https://github.com/viperproject/prusti-dev/blob/d5f1ac726df268dc71e4cd7235555985092ce0c3/prusti-viper/src/encoder/mir/specifications/specs.rs#L66)

Aurel300 commented on prusti-viper/src/encoder/mir/specifications/interface.rs:
```diff
@@ -1,39 +1,98 @@
-pub(crate) trait SpecificationsInterface {
-    fn is_pure(&self, def_id: DefId) -> bool;
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct SpecQuery<'tcx> {
```
Should this really be `pub`? Is it used outside of `specifications`?

vl0w replied:
Nope. I adjusted its visibility to `pub(in crate::encoder::mir::specifications)`

Aurel300 replied:
That's just `pub(super)`, no?

vl0w replied:
Hmm, yes 😅

Aurel300 commented on prusti-viper/src/encoder/mir/specifications/interface.rs:
```diff
@@ -1,39 +1,98 @@
+}
 
-    fn is_trusted(&self, def_id: DefId) -> bool;
+pub(crate) trait SpecificationsInterface<'tcx> {
```
General TODO (add in a comment, not for this PR): since the calls to `get_loop_specs` etc will receive `substs`, we could already monomorphise the MIR bodies.

Aurel300 commented on prusti-viper/src/encoder/encoder.rs:
```diff
@@ -543,7 +541,7 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
         debug!("encode_procedure({:?})", def_id);
         assert!(
-            !self.is_trusted(def_id),
+            !self.is_trusted( def_id, None),
```
```suggestion
            !self.is_trusted(def_id, None),
```

Aurel300 commented on prusti-viper/src/encoder/mir/specifications/constraints.rs:
```diff
@@ -0,0 +1,267 @@
+            SpecQueryCause::PureOrTrustedCheck => {
+                trace!(
+                "No need to resolve obligations for checking whether function is pure or trusted"
+            );
```
```suggestion
                    "No need to resolve obligations for checking whether function is pure or trusted"
                );
```

Aurel300 commented on prusti-viper/src/encoder/mir/specifications/constraints.rs:
```diff
@@ -0,0 +1,267 @@
+
+            // There is no caller when encoding a function.
+            // We still resolve obligations to account for constrained specs on a trait
+            // for which we encode it's implementation. The corresponding encoding will
```
```suggestion
            // for which we encode its implementation. The corresponding encoding will
```

Aurel300 commented on prusti-viper/src/encoder/mir/specifications/constraints.rs:
```diff
@@ -0,0 +1,267 @@
+mod resolvers {
+    use super::*;
+
+    pub mod trait_bounds {
```
Why these strange nested modules?

vl0w replied:
The idea was that we have one mod per `SpecConstraintKind`. I am not sure though whether we we will ever have multiple variants in there.
I will keep `SpecConstraintKind` but remove the outermost mod `resolvers` if thats ok for you. Keeping trait bound resolving logic in a separate module is fine to give `constraints.rs` some structure.

vl0w committed:
Typos, formatting

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

vl0w committed:
Merge branch 'master' into ghost-constraints

# Conflicts:
#	prusti-interface/src/specs/mod.rs
#	prusti-interface/src/specs/typed.rs
#	prusti-tests/tests/cargo_verify/prusti_toml/output.stdout
#	prusti-tests/tests/parse/ui/after_expiry.stdout
#	prusti-tests/tests/parse/ui/and.stdout
#	prusti-tests/tests/parse/ui/assert_on_expiry.stdout
#	prusti-tests/tests/parse/ui/composite.stdout
#	prusti-tests/tests/parse/ui/exists.stdout
#	prusti-tests/tests/parse/ui/expression.stdout
#	prusti-tests/tests/parse/ui/forall.stdout
#	prusti-tests/tests/parse/ui/implies.stdout
#	prusti-tests/tests/parse/ui/predicates-visibility.stdout
#	prusti-tests/tests/parse/ui/predicates.stdout
#	prusti-tests/tests/parse/ui/trait-bounds.stdout
#	prusti-tests/tests/parse/ui/traits.stdout
#	prusti-tests/tests/parse/ui/true.stdout
#	prusti-tests/tests/typecheck/ui/forall_encode_typeck.stdout
#	prusti-tests/tests/typecheck/ui/forall_triggers.stdout
#	prusti-tests/tests/typecheck/ui/nested_forall.stdout
#	prusti-tests/tests/verify/ui/calls.stdout
#	prusti-tests/tests/verify/ui/false.stdout
#	prusti-tests/tests/verify/ui/forall_verify.stdout
#	prusti-tests/tests/verify/ui/pledges.stdout
#	prusti-tests/tests/verify/ui/predicate.stdout
#	prusti-tests/tests/verify/ui/pure.stdout
#	prusti-viper/src/encoder/mir/specifications/interface.rs
#	prusti-viper/src/encoder/mir/specifications/specs.rs

vl0w committed:
Typos, formatting, docs

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

vl0w committed:
Remove test

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

vl0w committed:
Adjust visibility of SpecQuery

It should be only visible in prusti-viper::encoder::mir::specifications

vl0w committed:
Add doc comment for SpecGraph

vl0w committed:
Convert (ghost constraint) refinement tests to UI tests

vl0w committed:
Do not enable ghost constraints by default in test

+Adjust test to not check for unsoundness in ghost constraints

vl0w committed:
Undo weird parsing adjustment

vl0w committed:
Support generics and associated types in ghost constraints

Needed to support this
- Allow commas in ghost constraints during parsing
- The order of desugared generics in external trait specs has been adjusted (Prusti_T_Self must be at the beginning)

vl0w committed:
Merge branch 'master' into ghost-constraints

# Conflicts:
#	prusti-interface/src/specs/mod.rs
#	prusti-interface/src/specs/typed.rs
#	prusti-tests/tests/cargo_verify/prusti_toml/output.stdout
#	prusti-tests/tests/parse/ui/after_expiry.stdout
#	prusti-tests/tests/parse/ui/and.stdout
#	prusti-tests/tests/parse/ui/assert_on_expiry.stdout
#	prusti-tests/tests/parse/ui/composite.stdout
#	prusti-tests/tests/parse/ui/exists.stdout
#	prusti-tests/tests/parse/ui/expression.stdout
#	prusti-tests/tests/parse/ui/forall.stdout
#	prusti-tests/tests/parse/ui/implies.stdout
#	prusti-tests/tests/parse/ui/predicates-visibility.stdout
#	prusti-tests/tests/parse/ui/predicates.stdout
#	prusti-tests/tests/parse/ui/trait-bounds.stdout
#	prusti-tests/tests/parse/ui/traits.stdout
#	prusti-tests/tests/parse/ui/true.stdout
#	prusti-tests/tests/typecheck/ui/forall_encode_typeck.stdout
#	prusti-tests/tests/typecheck/ui/forall_triggers.stdout
#	prusti-tests/tests/typecheck/ui/nested_forall.stdout
#	prusti-tests/tests/verify/ui/calls.stdout
#	prusti-tests/tests/verify/ui/false.stdout
#	prusti-tests/tests/verify/ui/forall_verify.stdout
#	prusti-tests/tests/verify/ui/pledges.stdout
#	prusti-tests/tests/verify/ui/predicate.stdout
#	prusti-tests/tests/verify/ui/pure.stdout
#	prusti-viper/src/encoder/mir/pure/pure_functions/interpreter.rs
#	prusti-viper/src/encoder/mir/specifications/interface.rs
#	prusti-viper/src/encoder/mir/specifications/specs.rs

vl0w committed:
Post giga merge fixes

vl0w committed:
Fix/subst trait substs with call substs when fetching specs

vl0w committed:
Fix tests

vl0w committed:
UI tests are red.
Other tests are green.
Build is failing.
Jonas is maintaining.

vl0w committed:
Make SpecQuery an enum for less ambiguity

vl0w committed:
Merge branch 'vl0w-master' into ghost-constraints

# Conflicts:
#	prusti-contracts-impl/src/lib.rs
#	prusti-contracts-internal/src/lib.rs
#	prusti-contracts/src/lib.rs
#	prusti-interface/src/utils.rs
#	prusti-specs/src/lib.rs

vl0w committed:
Format, imports

Aurel300 commented on prusti-interface/src/specs/typed.rs:
```diff
@@ -1,70 +1,263 @@
+#[derive(Default, Debug, Clone)]
+pub struct DefSpecificationMap {
+    pub proc_specs: HashMap<DefId, SpecGraph<ProcedureSpecification>>,
+    pub loop_specs: HashMap<DefId, SpecGraph<LoopSpecification>>,
```
Does it make sense for this to be a `SpecGraph` as well? I guess for now there is no way to declare a ghost constrained loop invariant (if that even makes sense).

vl0w replied:
Nope, I think it's fine to omit the `SpecGraph` there

Aurel300 commented on prusti-interface/src/specs/typed.rs:
```diff
@@ -1,70 +1,263 @@
+/// Note: Unlike postconditions, the preconditions are not copied amongst the base spec
+/// and constrained specs.
+///
+/// # A note about behavior subtyping
```
`behavioural subtyping`

(or `behavioral` for AmEng 🤮)

vl0w replied:
My own variety of English is by far the most superior of all of them 😎

vl0w commented on prusti-viper/src/encoder/mir/specifications/specs.rs:
```diff
@@ -10,100 +16,169 @@ use prusti_interface::{
+
+                // We need to subst the trait substs because they might be generic
+                // (with substitutions in the call substs)
+                let trait_substs = if trait_substs.needs_subst() {
```
Why do we need to do this? Should this not be handled by `find_trait_method_substs`? @Aurel300

Aurel300 commented on prusti-specs/src/extern_spec_rewriter/common.rs:
```diff
@@ -142,11 +143,40 @@ impl<'a> AssociatedTypeRewriter<'a> {
+    fn visit_attribute_mut(&mut self, attr: &mut syn::Attribute) {
+        let self_replacement = self.self_type.to_token_stream().to_string();
+
+        // Note: Attribute tokens are not visited by a syn visitor,
+        // so we visit the attribute tokens "manually"
+        attr.tokens = self.rewrite_tokens(self_replacement.as_str(), attr.tokens.clone());
```
Does this mean we process here the user-written attribute for `Self` translation? This seems weird, I would expect us to do it after preparsing…

vl0w replied:
I think in order to do that we need to process the nested specs in an extern spec with `generate_spec_and_assertions` like we do for trait spec refinements.
I think this is generally a good idea.
Can we do this after we merged abstract predicates #938? I moved this logic to a new place and would like to avoid nasty merges. So my suggestion:
- Leave this as it is rn
- Finish abstract predicates
- New PR where we eagerly process nested specs in extern specs

Aurel300 commented on prusti-specs/src/extern_spec_rewriter/traits.rs:
```diff
@@ -184,6 +184,12 @@ impl<'a> GeneratedStruct<'a> {
+        // We need to rewrite `Self` in the attributes tokens as well
+        // to account for ghost constraints with a bound on `Self`
+        for attr in trait_method_attrs.iter_mut() {
+            rewriter.rewrite_attribute(attr);
```
Same comment as above.

Aurel300 commented on prusti-specs/src/ghost_constraints/mod.rs:
```diff
@@ -0,0 +1,142 @@
+            // Add attribute to mark this as a "specification with constraint" (used for processing the contract in `SpecCollector`)
+            item_fn.attrs.push(parse_quote_spanned!(tokens_span=>
+                #[prusti::ghost_constraint_trait_bounds_in_where_clause]
+            ));
```
I think it's cleaner/more consistent to use `{...}` for the `(parse_)quote(_spanned)!` macros.

Aurel300 commented on prusti-specs/src/ghost_constraints/mod.rs:
```diff
@@ -0,0 +1,142 @@
+    }
+}
+
+fn verify_provided_trait_bounds(trait_bounds: &ProvidedTraitBounds) -> syn::Result<()> {
```
In the context of a verifier, `verify` is a big word :D `validate`?

Aurel300 commented on prusti-specs/src/ghost_constraints/mod.rs:
```diff
@@ -0,0 +1,142 @@
+    Ok(())
+}
+
+fn verify_provided_trait_bound(trait_bound: &syn::TraitBound) -> syn::Result<()> {
```
Does this really need to be a method?

Aurel300 commented on prusti-specs/src/ghost_constraints/mod.rs:
```diff
@@ -0,0 +1,142 @@
+
+fn generate_where_clause_for_spec(
+    trait_bounds: &ProvidedTraitBounds,
+    preliminary_where_clause: Option<&syn::WhereClause>,
```
`preliminary` might be misleading, if this is referring to the `where` clause written outside a ghost constraint. `existing`?

Aurel300 commented on prusti-specs/src/ghost_constraints/mod.rs:
```diff
@@ -0,0 +1,142 @@
+                    loop {
+                        if input.is_empty() {
+                            break;
+                        }
```
```suggestion
                    while !input.is_empty() {
```

Aurel300 commented on prusti-specs/src/ghost_constraints/mod.rs:
```diff
@@ -0,0 +1,142 @@
+    predicate_type: syn::PredicateType,
+}
+
+impl syn::parse::Parse for ProvidedTraitBounds {
```
This parsing logic and the fact that we parse the bounds *at all* is only so that we can raise a parsing error for invalid bounds (e.g. lifetimes)?

vl0w replied:
I think it also comes in handy when generate the where clause for the spec function. Ghost constraint trait bounds are actually `syn::PredicateType`s, but they are not parsable by `syn` (see comment on the type). So I made this custom struct.

Do you think this could be done easier?

Aurel300 replied:
Well, if we simply passed through the predicate as typed by the user, without parsing it with `syn`, we wouldn't need this. But then some validation (e.g. that lifetime bounds aren't allowed) would need to happen post-typing and maybe that's too much hassle.

Aurel300 commented on prusti-specs/src/ghost_constraints/mod.rs:
```diff
@@ -0,0 +1,142 @@
+                            break;
+                        }
+
+                        let punct = input.parse()?;
```
Why the `value` and `punct` temporaries?

Aurel300 commented on prusti-specs/src/ghost_constraints/mod.rs:
```diff
@@ -0,0 +1,142 @@
+                bounded_ty: input.parse()?,
+                colon_token: input.parse()?,
+                bounds: {
+                    let mut bounds = syn::punctuated::Punctuated::new();
```
I'm surprised that this is implemented with a loop: does [`Punctuated::parse_separated_nonempty`](https://docs.rs/syn/latest/syn/punctuated/struct.Punctuated.html#method.parse_separated_nonempty) not do the same thing?

vl0w replied:
I think it does! :)

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -50,6 +50,12 @@ pub fn parse_prusti_assert_pledge(tokens: TokenStream) -> syn::Result<(TokenStre
+pub fn parse_ghost_constraint(tokens: TokenStream) -> syn::Result<(TokenStream, Vec<NestedSpec<TokenStream>>)> {
+    let pts = PrustiTokenStream::new(tokens);
+    let (trait_bounds_ts, nested_specs_ts) = pts.parse_ghost_constraint()?;
+    Ok((trait_bounds_ts, nested_specs_ts))
```
```suggestion
    pts.parse_ghost_constraint()
```

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -147,7 +155,20 @@ impl PrustiTokenStream {
+        for token in other.tokens.into_iter() {
+            self.tokens.push_back(token)
+        }
     }
```
I don't think these should need to exist…

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -214,6 +235,44 @@ impl PrustiTokenStream {
+            if arg_iter.peek().is_some() {
+                // arg is part of ghost constraint
+                trait_bounds_ts.extend(arg);
+                trait_bounds_ts.tokens.push_back(PrustiToken::Token(TokenTree::Punct(Punct::new(',', Spacing::Alone))))
```
Why are you extending a `PrustiTokenStream`? Why not just a `TokenStream`?

vl0w replied:
Because the ghost constraints are `PrustiTokenStream`s (PTS) and I first want to collect all ghost constraint into one single `PrustiTokenStream` before I can finally parse it with `parse_rust_only`.
When having something like `#[ghost_constraint(T: A<Foo, Baz = Bar>, ...)]`, there will be 3 PTS:
- (1): `T: A<Foo`
- (2): ` Baz = Bar>`
- (3): `...`
I think need to combine (1) and (2) into a new PTS first such that I do not get into unnecessary parsing errors.
Do you think I should immediately call `parse_rust_only` on (1) and (2) and combine the corresponding `TokenStream`s afterwards?

Aurel300 replied:
Yes, I think parsing as "Rust only" and pushing into a `TokenStream` is nicer than extending the `PrustiTokenStream` API to support this.

vl0w replied:
Done

Aurel300 commented on prusti-tests/tests/verify/ui/ghost-constraints/invalid-trait-refinement-2.rs:
```diff
@@ -0,0 +1,49 @@
+    fn foo(&mut self);
+}
+
+#[extern_spec] //~ ERROR: the method's postcondition may not be a valid strengthening of the trait's postcondition.
```
Does the `//~ ERROR` annotation do anything in UI tests? (Also in the `invalid-trait-refinement-1.rs` file.)

vl0w replied:
No, I think this is a leftover, they once were not UI tests

Aurel300 commented on prusti-tests/tests/verify_overflow/pass/ghost-constraints/split-constraint.rs:
```diff
@@ -0,0 +1,24 @@
+#[ghost_constraint(T: A, [
+    ensures(result % 2 == 0)
+])]
+#[ghost_constraint(T: A, [
```
Is this meant to be allowed? Is it a syntactic check for whether the bound is the same?

vl0w replied:
This is just a check that you indeed can have two constraints IF the bounds are equal.
In this case, two spec functions with a where clause `where T: A` will be generated.
When I resolve the constraints, I first collect all of the spec function's `ParamEnv`s with their included where clauses and then compare them. If not equal => throw error. But they will be equal in that case, so no need to be restrictive there.

Aurel300 commented on prusti-viper/src/encoder/mir/specifications/constraints.rs:
```diff
@@ -0,0 +1,273 @@
+        } else {
+            trace!("Constraint not fulfilled");
+            false
+        }
```
```suggestion
        trace!("Constraint fulfilled: {all_bounds_satisfied}");
        all_bounds_satisfied
```

Aurel300 commented on prusti-viper/src/encoder/mir/specifications/constraints.rs:
```diff
@@ -0,0 +1,273 @@
+        let param_env_lookup = if let Some(caller_def_id) = context.caller_proc_def_id {
+            env.tcx().param_env(caller_def_id)
+        } else {
+            ty::ParamEnv::reveal_all()
```
Should this not be the `param_env` of the method itself?

Aurel300 commented on prusti-interface/src/environment/mod.rs:
```diff
@@ -534,4 +540,45 @@ impl<'tcx> Environment<'tcx> {
+            param_env,
+            recursion_depth: 0, // Arbitrarily chosen magic number
+            predicate,
+        };
```
```suggestion
        let obligation = traits::Obligation::new(
            traits::ObligationCause::dummy(),
            param_env,
            predicate,
        );
```

(also gets rid of the not-quite-magic magic)

Aurel300 commented on prusti-interface/src/environment/mod.rs:
```diff
@@ -534,4 +540,45 @@ impl<'tcx> Environment<'tcx> {
+                normalizable,
+                &mut obligations,
+            )
+        })
```
Do our tests pass if we `assert_eq!(obligations.len(), 0);` here?

vl0w replied:
No, sadly not. And it might be a problem.
Take the test `associated_types_3.rs`.
When we encode the function `Foo::foo`, we resolve the ghost constraints, but the associated type projection can not be fully normalized. This means that we do no really consider the constrained spec in that case (=> behavioral subtyping will also not be checked).

vl0w committed:
Small changes from review

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

vl0w committed:
Loop specs are not stored inside a SpecGraph

vl0w committed:
Clean ui tests

vl0w committed:
Use param env of procedure to be encoded when resolving ghost constraints

vl0w committed:
Refactor ghost constraint parsing

vl0w committed:
Substitute adapted trait substs with call substs

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

vl0w committed:
Merge branch 'vl0w-master' into ghost-constraints

vl0w committed:
Pass call substs to is_pure call
